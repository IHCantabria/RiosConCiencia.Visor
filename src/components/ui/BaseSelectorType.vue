<template>
  <div class="inner-container">
    <span v-if="isEmptyList" class="empty-list">No Options available</span>
    <vs-select
      v-else
      v-model="activeItem"
      :label="label"
      :width="widthSelect"
      autocomplete
    >
      <vs-select-item
        v-for="(item, index) in items"
        :key="index"
        class="selectStyle"
        :value="item"
        :text="item.name"
      />
    </vs-select>
  </div>
</template>
<script>
export default {
  props: {
    items: {
      type: Array,
      required: true
    },
    label: {
      type: String,
      required: false,
      default: () => {
        return "";
      }
    },
    widthSelect: {
      type: String,
      required: false,
      default: () => {
        return "400px";
      }
    },
    lastActiveItem: {
      type: Object,
      required: false,
      default: () => {
        return null;
      }
    }
  },
  data() {
    return {
      activeItem: {}
    };
  },
  computed: {
    isEmptyList() {
      return this.items.length === 0;
    }
  },
  watch: {
    items: {
      handler() {
        this.selectFirstOption();
      },
      immediate: true
    },
    activeItem: {
      handler() {
        if (this.activeItem) {
          this.$emit("change", this.activeItem);
        }
      },
      deep: true
    }
  },
  mounted() {
    this.activeItem = this.lastActiveItem ? this.lastActiveItem : null;
    if (!this.activeItem) this.selectFirstOption();
  },
  methods: {
    selectFirstOption() {
      this.$nextTick(() => {
        this.activeItem = this.items[0]; // default selection
      }); // default selection
    }
  }
};
</script>
<style lang="scss" scoped>
.inner-container {
  display: flex;
  flex-grow: 1;
  justify-content: flex-start;
  align-items: center;
}

.selectStyle {
  margin: 10px;
}
</style>
