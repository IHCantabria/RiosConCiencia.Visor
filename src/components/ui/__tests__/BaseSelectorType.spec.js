import { shallowMount, createLocalVue } from "@vue/test-utils";
import BaseSelectorType from "../BaseSelectorType.vue";
import Vuesax from "vuesax";
import merge from "lodash.merge";

const localVue = createLocalVue();
localVue.use(Vuesax);

function createWrapper(overrides) {
  const defaultMountingOptions = {
    propsData: {
      items: [
        {
          id: 1,
          name: "Mareógrafo"
        },
        {
          id: 2,
          name: "Boya de costa"
        },
        {
          id: 3,
          name: "Boya Station"
        },
        {
          id: 4,
          name: "Restringida"
        },
        {
          id: 5,
          name: "Virtual Hindcast"
        }
      ]
    },
    localVue
  };
  return shallowMount(
    BaseSelectorType,
    merge(defaultMountingOptions, overrides)
  );
}

describe("baseSelectorType.vue", () => {
  //UNIT TESTS
  it("computed isEmptyList updated when items length change", () => {
    expect.assertions(2);
    const wrapper = createWrapper();
    expect(wrapper.vm.isEmptyList).toBe(false);
    wrapper.setProps({ items: [] });
    expect(wrapper.vm.isEmptyList).toBe(true);
  });
  it("change is emitted when watcher was actived", () => {
    expect.assertions(1);
    const wrapper = createWrapper();
    wrapper.setData({ activeItem: {} });
    expect(wrapper.emitted().change).toStrictEqual([[{}]]);
  });
  it("activeItem is defined when mounted is called", () => {
    expect.assertions(1);
    const wrapper = createWrapper({
      propsData: {
        lastActiveItem: {}
      }
    });
    expect(wrapper.vm.activeItem).toStrictEqual({});
  });
  it("selectFirstOption is called after mounted if activeItem is null", () => {
    expect.assertions(1);
    const selectFirstOption = jest.fn();
    const wrapper = createWrapper({
      propsData: {
        lastActiveItem: null
      },
      methods: {
        selectFirstOption
      }
    });
    expect(wrapper.vm.activeItem).toBeNull();
  });
  it("selectFirstOption method set activeItem with the first element of items", () => {
    expect.assertions(1);
    const wrapper = createWrapper({
      propsData: {
        lastActiveItem: {
          id: 1,
          name: "Mareógrafo"
        }
      }
    });
    expect(wrapper.vm.activeItem).toStrictEqual(wrapper.vm.items[0]);
  });
  //SNAPSHOT TESTS
  it("renders correctly with the default values", () => {
    expect.assertions(1);
    const wrapper = createWrapper();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("renders correctly with custom widthSelect", () => {
    expect.assertions(1);
    const wrapper = createWrapper({
      propsData: {
        widthSelect: "800px"
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("renders correctly with label", () => {
    expect.assertions(1);
    const wrapper = createWrapper({
      propsData: {
        label: "Boyas"
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("renders correctly with lastActiveItem selected", () => {
    expect.assertions(1);
    const wrapper = createWrapper({
      propsData: {
        lastActiveItem: {
          id: 5,
          name: "Virtual Hindcast"
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
