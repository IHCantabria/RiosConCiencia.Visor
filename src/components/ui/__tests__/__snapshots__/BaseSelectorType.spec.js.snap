// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`baseSelectorType.vue renders correctly with custom widthSelect 1`] = `
<div
  class="inner-container"
>
  <vs-select-stub
    autocomplete="true"
    color="primary"
    icon="keyboard_arrow_down"
    iconclear="close"
    iconpack="material-icons"
    label=""
    nodata="data no available"
    width="800px"
  >
    <vs-select-item-stub
      class="selectStyle"
      text="Mare贸grafo"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Boya de costa"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Boya Station"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Restringida"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Virtual Hindcast"
      value="[object Object]"
    />
  </vs-select-stub>
</div>
`;

exports[`baseSelectorType.vue renders correctly with label 1`] = `
<div
  class="inner-container"
>
  <vs-select-stub
    autocomplete="true"
    color="primary"
    icon="keyboard_arrow_down"
    iconclear="close"
    iconpack="material-icons"
    label="Boyas"
    nodata="data no available"
    width="400px"
  >
    <vs-select-item-stub
      class="selectStyle"
      text="Mare贸grafo"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Boya de costa"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Boya Station"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Restringida"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Virtual Hindcast"
      value="[object Object]"
    />
  </vs-select-stub>
</div>
`;

exports[`baseSelectorType.vue renders correctly with lastActiveItem selected 1`] = `
<div
  class="inner-container"
>
  <vs-select-stub
    autocomplete="true"
    color="primary"
    icon="keyboard_arrow_down"
    iconclear="close"
    iconpack="material-icons"
    label=""
    nodata="data no available"
    value="[object Object]"
    width="400px"
  >
    <vs-select-item-stub
      class="selectStyle"
      text="Mare贸grafo"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Boya de costa"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Boya Station"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Restringida"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Virtual Hindcast"
      value="[object Object]"
    />
  </vs-select-stub>
</div>
`;

exports[`baseSelectorType.vue renders correctly with the default values 1`] = `
<div
  class="inner-container"
>
  <vs-select-stub
    autocomplete="true"
    color="primary"
    icon="keyboard_arrow_down"
    iconclear="close"
    iconpack="material-icons"
    label=""
    nodata="data no available"
    width="400px"
  >
    <vs-select-item-stub
      class="selectStyle"
      text="Mare贸grafo"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Boya de costa"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Boya Station"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Restringida"
      value="[object Object]"
    />
    <vs-select-item-stub
      class="selectStyle"
      text="Virtual Hindcast"
      value="[object Object]"
    />
  </vs-select-stub>
</div>
`;
